File: sdk/typescript/src/server/groups.ts
import { listQuery, type HttpTransport } from "../http.js";
import type { ListParams, Membership, Group } from "../types.js";

export class ServerGroupsService {
  constructor(private readonly http: HttpTransport) {}

  async create(input: { name: string; permissions?: string[] }): Promise<Group> {
    return this.http.request<Group>("POST", "/v1/server/groups", {
      auth: "apiKey",
      body: input,
    });
  }

  async list(params?: ListParams): Promise<Group[]> {
    const res = await this.http.request<{ groups: Group[] }>("GET", "/v1/server/groups", {
      auth: "apiKey",
      query: listQuery(params),
    });
    return res.groups ?? [];
  }

  async get(id: string): Promise<Group> {
    return this.http.request<Group>("GET", `/v1/server/groups/${id}`, { auth: "apiKey" });
  }

  async getPrefs(id: string): Promise<Record<string, unknown>> {
    const res = await this.http.request<{ prefs: Record<string, unknown> }>(
      "GET",
      `/v1/server/groups/${id}/prefs`,
      { auth: "apiKey" }
    );
    return res.prefs ?? {};
  }

  async updatePrefs(id: string, prefs: Record<string, unknown>): Promise<Record<string, unknown>> {
    const res = await this.http.request<{ prefs: Record<string, unknown> }>(
      "PUT",
      `/v1/server/groups/${id}/prefs`,
      { auth: "apiKey", body: { id, prefs } }
    );
    return res.prefs ?? {};
  }

  async delete(id: string): Promise<void> {
    await this.http.request<void>("DELETE", `/v1/server/groups/${id}`, { auth: "apiKey" });
  }

  async createMembership(
    groupId: string,
    input: {
      email?: string;
      user_id?: string;
      name?: string;
      roles?: string[];
      status?: string;
    }
  ): Promise<Membership> {
    return this.http.request<Membership>("POST", `/v1/server/groups/${groupId}/memberships`, {
      auth: "apiKey",
      body: { group_id: groupId, ...input },
    });
  }

  async listMemberships(groupId: string, params?: ListParams): Promise<Membership[]> {
    const res = await this.http.request<{ memberships: Membership[] }>(
      "GET",
      `/v1/server/groups/${groupId}/memberships`,
      { auth: "apiKey", query: listQuery(params) }
    );
    return res.memberships ?? [];
  }

  async getMembership(groupId: string, membershipId: string): Promise<Membership> {
    return this.http.request<Membership>(
      "GET",
      `/v1/server/groups/${groupId}/memberships/${membershipId}`,
      { auth: "apiKey" }
    );
  }

  async updateMembership(
    groupId: string,
    membershipId: string,
    roles: string[]
  ): Promise<Membership> {
    return this.http.request<Membership>(
      "PATCH",
      `/v1/server/groups/${groupId}/memberships/${membershipId}`,
      {
        auth: "apiKey",
        body: { group_id: groupId, membership_id: membershipId, roles },
      }
    );
  }

  async updateMembershipStatus(
    groupId: string,
    membershipId: string,
    status: string
  ): Promise<Membership> {
    return this.http.request<Membership>(
      "PATCH",
      `/v1/server/groups/${groupId}/memberships/${membershipId}/status`,
      {
        auth: "apiKey",
        body: { group_id: groupId, membership_id: membershipId, status },
      }
    );
  }

  async deleteMembership(groupId: string, membershipId: string): Promise<void> {
    await this.http.request<void>(
      "DELETE",
      `/v1/server/groups/${groupId}/memberships/${membershipId}`,
      { auth: "apiKey" }
    );
  }
}

File: sdk/go/server/groups.go
package server

import (
	"context"

	serverv1 "github.com/torchwooddev/torchwood/genproto/server/v1"
	sharedv1 "github.com/torchwooddev/torchwood/genproto/shared/v1"
)

// GroupsService 封装 Server API 的用户组管理服务。
type GroupsService struct{ c *Client }

// CreateGroup 创建用户组（可附带权限）。
func (t *GroupsService) CreateGroup(ctx context.Context, name string, permissions []string) (*serverv1.Group, error) {
	return t.c.groups.CreateGroup(ctx, &serverv1.CreateGroupRequest{
		Name:        name,
		Permissions: permissions,
	})
}

// GetGroup 按 ID 获取用户组。
func (t *GroupsService) GetGroup(ctx context.Context, groupID string) (*serverv1.Group, error) {
	return t.c.groups.GetGroup(ctx, &serverv1.GetGroupRequest{Id: groupID})
}

// DeleteGroup 删除用户组。
func (t *GroupsService) DeleteGroup(ctx context.Context, groupID string) error {
	_, err := t.c.groups.DeleteGroup(ctx, &serverv1.GetGroupRequest{Id: groupID})
	return err
}

// GetGroupPrefs 获取用户组偏好。
func (t *GroupsService) GetGroupPrefs(ctx context.Context, groupID string) (*serverv1.GetGroupPrefsResponse, error) {
	return t.c.groups.GetGroupPrefs(ctx, &serverv1.GetGroupRequest{Id: groupID})
}

// UpdateGroupPrefs 全量替换用户组偏好。
func (t *GroupsService) UpdateGroupPrefs(ctx context.Context, groupID string, prefs map[string]any) (*serverv1.GetGroupPrefsResponse, error) {
	prefsStruct, err := toStruct(prefs)
	if err != nil {
		return nil, err
	}
	return t.c.groups.UpdateGroupPrefs(ctx, &serverv1.UpdateGroupPrefsRequest{
		Id:    groupID,
		Prefs: prefsStruct,
	})
}

// ListGroups 按查询 DSL 列出用户组。
func (t *GroupsService) ListGroups(ctx context.Context, queries []string, pageSize int32, pageToken string) (*serverv1.ListGroupsResponse, error) {
	return t.c.groups.ListGroups(ctx, &sharedv1.ListRequest{
		PageSize:  pageSize,
		PageToken: pageToken,
		Queries:   queries,
	})
}

// CreateMembership 创建用户组成员关系（按 userID 或邮箱）。
func (t *GroupsService) CreateMembership(ctx context.Context, groupID, userID, email, name string, roles []string, status string) (*serverv1.Membership, error) {
	return t.c.groups.CreateMembership(ctx, &serverv1.CreateMembershipRequest{
		GroupId: groupID,
		UserId:  userID,
		Email:   email,
		Name:    name,
		Roles:   roles,
		Status:  status,
	})
}

// ListMemberships 列出用户组成员。
func (t *GroupsService) ListMemberships(ctx context.Context, groupID string) (*serverv1.ListMembershipsResponse, error) {
	return t.c.groups.ListMemberships(ctx, &serverv1.ListMembershipsRequest{GroupId: groupID})
}

// GetMembership 获取成员关系。
func (t *GroupsService) GetMembership(ctx context.Context, groupID, membershipID string) (*serverv1.Membership, error) {
	return t.c.groups.GetMembership(ctx, &serverv1.GetMembershipRequest{
		GroupId:      groupID,
		MembershipId: membershipID,
	})
}

// UpdateMembership 更新成员角色。
func (t *GroupsService) UpdateMembership(ctx context.Context, groupID, membershipID string, roles []string) (*serverv1.Membership, error) {
	return t.c.groups.UpdateMembership(ctx, &serverv1.UpdateMembershipRequest{
		GroupId:      groupID,
		MembershipId: membershipID,
		Roles:        roles,
	})
}

// UpdateMembershipStatus 更新成员状态。
func (t *GroupsService) UpdateMembershipStatus(ctx context.Context, groupID, membershipID, status string) (*serverv1.Membership, error) {
	return t.c.groups.UpdateMembershipStatus(ctx, &serverv1.UpdateMembershipStatusRequest{
		GroupId:      groupID,
		MembershipId: membershipID,
		Status:       status,
	})
}

// DeleteMembership 移除成员关系。
func (t *GroupsService) DeleteMembership(ctx context.Context, groupID, membershipID string) error {
	_, err := t.c.groups.DeleteMembership(ctx, &serverv1.GetMembershipRequest{
		GroupId:      groupID,
		MembershipId: membershipID,
	})
	return err
}
